x/bank/src/keeper/bank.rs

```diff
@@ -0,0 +1,131 @@
+use std::{collections::HashSet, sync::OnceLock};
+
+use super::*;
+
+impl<
+        SK: StoreKey,
+        PSK: ParamsSubspaceKey,
+        AK: AuthKeeper<SK, M> + Send + Sync + 'static,
+        M: Module + strum::IntoEnumIterator,
+    > BankKeeper<SK, M> for Keeper<SK, PSK, AK, M>
+{
+    fn send_coins_from_account_to_module<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        from_address: AccAddress,
+        to_module: &M,
+        amount: UnsignedCoins,
+    ) -> Result<(), BankKeeperError> {
+        self.auth_keeper
+            .check_create_new_module_account(ctx, to_module)?;
+
+        let msg = MsgSend {
+            from_address,
+            to_address: to_module.address(),
+            amount,
+        };
+
+        self.send_coins(ctx, msg)?;
+
+        Ok(())
+    }
+
+    fn denom_metadata<DB: Database, CTX: QueryableContext<DB, SK>>(
+        &self,
+        ctx: &CTX,
+        base: &Denom,
+    ) -> Result<Option<Metadata>, GasStoreErrors> {
+        let bank_store = ctx.kv_store(&self.store_key);
+        let denom_metadata_store = bank_store.prefix_store(denom_metadata_key(base.to_string()));
+
+        Ok(denom_metadata_store
+            .get(&base.to_string().into_bytes())?
+            .map(|metadata| {
+                Metadata::decode::<&[u8]>(&metadata)
+                    .ok()
+                    .unwrap_or_corrupt()
+            }))
+    }
+
+    fn coins_burn<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        module: &M,
+        deposit: &UnsignedCoins,
+    ) -> Result<(), BankKeeperError> {
+        let module_acc_addr = module.address();
+
+        let account = self
+            .auth_keeper
+            .get_account(ctx, &module_acc_addr)?
+            .ok_or(AccountNotFound::new(module_acc_addr.to_string()))?;
+
+        match account.has_permissions("burner") {
+            true => Ok(()),
+            false => Err(BankKeeperError::AccountPermission),
+        }?;
+
+        self.sub_unlocked_coins(ctx, &module_acc_addr, deposit)?;
+
+        for coin in deposit.inner() {
+            let supply = self.supply(ctx, &coin.denom)?;
+            // we slightly different from cosmos in this. They return coin always.
+            // If no value found then new coin with zero balance, subtraction and call of set_supply which deletes coins with zero balance
+            // We omitted it but if any issue arise be aware that maybe we should delete zero coins if we store any.
+            if let Some(mut supply) = supply {
+                supply.amount.sub_assign(coin.amount);
```

Comment:
Does this panic if supply < coin.amount?

Reply:
This type from `cosmwasms-std`. I think it follows standart logic for rust. panic for debug overflow for release